src/persist/src/operators/stream.rs

```diff
@@ -198,55 +200,78 @@ where
         // capabilities.
         let active_seal_operator = self.scope().index() == 0;
 
-        seal_op.build(move |mut capabilities| {
-            let mut err_cap = Some(capabilities.pop().expect("missing capability"));
-            let mut _data_cap = capabilities.pop();
-
+        seal_op.build(move |_capabilities| {
             move |frontiers| {
                 let mut data_output = data_output.activate();
                 let mut error_output = error_output.activate();
 
                 // Pass through all data.
-                data_input.for_each(|time, data| {
+                data_input.for_each(|cap, data| {
                     data.swap(&mut data_buffer);
 
-                    let mut session = data_output.session(&time);
+                    let mut session = data_output.session(&cap);
                     session.give_vec(&mut data_buffer);
-                });
 
-                // Seal if/when the frontier advances.
-                let new_input_frontier = frontiers[0].frontier();
-                let progress =
-                    !PartialOrder::less_equal(&new_input_frontier, &input_frontier.borrow());
+                    // We only need capabilities for reporting errors, which we only need to do
+                    // when we're the active operator.
+                    if active_seal_operator {
+                        capabilities.insert(cap.retain_for_output(1));
+                    }
+                });
 
-                if progress {
-                    input_frontier.clear();
-                    input_frontier.extend(new_input_frontier.into_iter().cloned());
+                if active_seal_operator {
+                    // Seal if/when the frontier advances.
+                    let new_input_frontier = frontiers[0].frontier();
+                    let progress =
+                        !PartialOrder::less_equal(&new_input_frontier, &input_frontier.borrow());
+
+                    if progress {
+                        input_frontier.clear();
+                        input_frontier.extend(new_input_frontier.into_iter().cloned());
```

Comment:
something something clone_from? (this is a shot in the dark, I'm not sure if it's relevant here)

Reply:
It's a good shot in the dark, but the types don't line up. And this idiom is also used elsewhere in the code, when managing input frontiers.